kernel/kheap.c

```diff
@@ -77,9 +77,8 @@ void *k_heap_aligned_alloc(struct k_heap *h, size_t align, size_t bytes,
 	while (ret == NULL) {
 		ret = sys_heap_aligned_alloc(&h->heap, align, bytes);
 
-		now = sys_clock_tick_get();
 		if (!IS_ENABLED(CONFIG_MULTITHREADING) ||
-		    (ret != NULL) || ((end - now) <= 0)) {
+		    (ret != NULL) || K_TIMEOUT_EQ(timeout, K_NO_WAIT)) {
```

Comment:
This bit seems unrelated and AFAICT is a noop.  There is K_NO_WAIT handling in sys_clock_timeout_end_calc() already.

Reply:
Oh, actually this is a bug.  You can have "end < now" as a perfectly valid state, either because of a race between the two tick calls involved or because the upstream caller passed an absolute timeout in the past.  This code will then pend incorrectly in that case.